File: 2021/3/main.py
s = [i.strip() for i in open("input.txt").read().split()]
from collections import defaultdict
from itertools import filterfalse
d = defaultdict(int)
for i in s:
    for j in range(len(i)):
        d[j] += int(i[j])
x = ""
for i in d:
    if d[i] > len(s) // 2:
        x += "1"
    else:
        x += "0"

y = "".join("1" if i == "0" else "0" for i in x)
print(int(x, 2)*int(y, 2))

x = s[:]
z = 0
while len(x) > 1:
    y = sum(int(i[z]) for i in x)
    if y >= len(x) - y:
        x[:] = filterfalse(lambda x: x[z] == "0", x)
    else:
        x[:] = filterfalse(lambda x: x[z] == "1", x)
    z += 1
a = x[0]

x = s[:]
z = 0
while len(x) > 1:
    y = sum(int(i[z]) for i in x)
    if y < len(x) - y:
        x[:] = filterfalse(lambda x: x[z] == "0", x)
    else:
        x[:] = filterfalse(lambda x: x[z] == "1", x)
    z += 1
b = x[0]
print(int(a, 2) * int(b, 2))

File: 2021/3/main.jl
using DataStructures

function p1()
    d = DataStructures.DefaultDict{Int, Int}(0)
    len = 12
    for line in readlines("input.txt")
        for i in 1:len
            d[i] += line[i] == '1' ? 1 : -1
        end
    end
    gamma, epsilon = 0, 0
    for i in 1:len
        if d[i] > 0
            gamma += 2 ^ (len - i)
        else
            epsilon += 2 ^ (len - i)
        end
    end
    println(gamma * epsilon)
end

function p2()
    s1 = Set{String}(readlines("input.txt"))
    s2 = Set{String}(readlines("input.txt"))
    i = 1
    while length(s1) != 1
        new = Set{String}()
        if sum(x -> x[i] == '1', s1) >= length(s1) / 2
            for s in s1
                if s[i] == '1'
                    push!(new, s)
                end
            end
        else
            for s in s1
                if s[i] == '0'
                    push!(new, s)
                end
            end
        end
        i += 1
        empty!(s1)
        union!(s1, new)
    end
    i = 1
    while length(s2) != 1
        new = Set{String}()
        if sum(x -> x[i] == '1', s2) < length(s2) / 2
            for s in s2
                if s[i] == '1'
                    push!(new, s)
                end
            end
        else
            for s in s2
                if s[i] == '0'
                    push!(new, s)
                end
            end
        end
        i += 1
        empty!(s2)
        union!(s2, new)
    end
    println(parse(Int, pop!(s1); base=2) * parse(Int, pop!(s2); base=2))
end

p1()
p2()
